demos/programs/popups/autopopups.man
.\" $XConsortium: autopopups.man /main/4 1995/07/17 10:48:39 drk $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY

.TH autopopups 1X MOTIF "Demonstration programs"
.SH NAME
\*Lautopopups\*O\(emAutomatic popup menu demonstration
.SH SYNOPSIS
.sS
\*Lautopopups\*O
.sE
.SH DESCRIPTION
\*Lautopopups\*O
demonstrates the use of the new automatic popup features in 2.0.  Note
the different popup menus which are posted as you press button3 on the 
different objects.  Press the Quit button to exit.



.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


demos/programs/popups/autopopups.c
/* $XConsortium: autopopups.c /main/5 1995/07/15 20:45:56 drk $ */
/*
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 */
/*
 * HISTORY
 */

#include <stdio.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/IntrinsicP.h>
#include <X11/CompositeP.h>
#include <Xm/XmP.h>
#include <Xm/Frame.h>
#include <Xm/Label.h>
#include <Xm/PushB.h>
#include <Xm/PushBG.h>
#include <Xm/LabelG.h>
#include <Xm/SeparatoG.h>
#include <Xm/RowColumn.h>
#include <Xm/BulletinB.h>
#include <Xm/DrawingA.h>

XtAppContext context;
#define APP_CLASS "AutomaticPopups"

Widget exitpb;

void QuitAppl(Widget, char *, XmPushButtonCallbackStruct *);
void ExposeCB(Widget, char *, XmDrawingAreaCallbackStruct *);
void ButtonEventHandler(Widget, XtPointer, XEvent*);
void PopupHandlerCB(Widget, XtPointer, XmPopupHandlerCallbackStruct *);
void addButton(Widget, char*);
void addChildren(Widget, char*);

int circlex, circley, circled;
int rectx, recty, recth, rectw;

Widget	popup1, popup2, popup3, popup4, popup5;

int main(int argc, char **argv)
{
  Widget        shell;
  Widget        frame1, frame2;
  Widget	da1, da2;
  Widget	btn1, btn2, label1, label2, btn3, sbtn1;
  Widget	explain;
  Arg		args[10];
  int		count;
  XmString	str, tmp1, tmp2;

  shell     = XtVaAppInitialize(&context, APP_CLASS, NULL, 0, &argc, argv,
				NULL, NULL);

  frame1 = XmCreateFrame(shell, "outerframe", NULL, 0);
  XtManageChild(frame1);

  count = 0;
  XtSetArg(args[count], XmNpopupEnabled, XmPOPUP_AUTOMATIC_RECURSIVE); count++;
  popup1 = XmCreatePopupMenu(frame1, "popupOF", args, count);

  addChildren(popup1, "OuterFrameMenu");

  str = XmStringCreateLocalized("OuterFrame");
  count = 0;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNframeChildType, XmFRAME_TITLE_CHILD); count++;
  label1 = XmCreateLabel(frame1, "label1", args, count);
  XtManageChild(label1);
  XmStringFree(str);

  count = 0;
  XtSetArg(args[count], XmNheight, 500); count++;
  XtSetArg(args[count], XmNwidth, 500); count++;
  XtSetArg(args[count], XmNframeChildType, XmFRAME_WORKAREA_CHILD); count++;
  XtSetArg(args[count], XmNresizePolicy, XmRESIZE_NONE); count++;
  da1 = XmCreateDrawingArea(frame1, "da1", args, count);
  XtManageChild(da1);

  count = 0;
  XtSetArg(args[count], XmNx, 100); count++;
  XtSetArg(args[count], XmNy, 100); count++;
  frame2 = XmCreateFrame(da1, "innerframe", args, count);
  XtManageChild(frame2);

  count = 0;
  XtSetArg(args[count], XmNpopupEnabled, XmPOPUP_AUTOMATIC_RECURSIVE); count++;
  popup2 = XmCreatePopupMenu(frame2, "popupIF", args, count);

  addChildren(popup2, "InnerFrameMenu");
  addButton(popup2, "Make me different");

  str = XmStringCreateLocalized("InnerFrame");
  count = 0;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNframeChildType, XmFRAME_TITLE_CHILD); count++;
  label2 = XmCreateLabel(frame2, "label2", args, count);
  XtManageChild(label2);
  XmStringFree(str);

  count = 0;
  XtSetArg(args[count], XmNheight, 300); count++;
  XtSetArg(args[count], XmNwidth, 300); count++;
  XtSetArg(args[count], XmNframeChildType, XmFRAME_WORKAREA_CHILD); count++;
  XtSetArg(args[count], XmNresizePolicy, XmRESIZE_NONE); count++;
  da2 = XmCreateDrawingArea(frame2, "da2", args, count);
  XtManageChild(da2);
  XtAddCallback(da2, XmNexposeCallback, (XtCallbackProc) ExposeCB, NULL);
  XtAddCallback(da2, XmNpopupHandlerCallback, 
		(XtCallbackProc) PopupHandlerCB, NULL);

  count = 0;
  str = XmStringCreateLocalized(
	"Click Button3 anywhere to cause a popup menu");
  tmp1 = XmStringConcat(str, XmStringSeparatorCreate());
  XmStringFree(str);
  tmp2 = XmStringCreateLocalized(
         "to be posted.  Note that the posted menu");
  str = XmStringConcat(tmp1, tmp2);
  XmStringFree(tmp1);
  XmStringFree(tmp2);
  tmp1 = XmStringConcat(str, XmStringSeparatorCreate());
  XmStringFree(str);
  tmp2 = XmStringCreateLocalized(
         "differs depending on where you click.");
  str = XmStringConcat(tmp1, tmp2);
  XmStringFree(tmp1);
  XmStringFree(tmp2);

  XtSetArg(args[count], XmNx, 10); count++;
  XtSetArg(args[count], XmNy, 200); count++;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNalignment, XmALIGNMENT_BEGINNING); count++;
  explain = XmCreateLabelGadget(da2, "Explain", args, count);
  XtManageChild(explain);
  XmStringFree(str);

  circlex = 40; circley = 120; circled = 30;
  rectx = 80; recty = 120; recth = 30; rectw = 60;

  str = XmStringCreateLocalized("Quit");
  count = 0;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNx, 40); count++;
  XtSetArg(args[count], XmNy, 40); count++;
  btn1 = XmCreatePushButton(da2, "pb1", args, count);
  XtManageChild(btn1);
  XtAddCallback(btn1, XmNactivateCallback, 
		(XtCallbackProc) QuitAppl, NULL);

  str = XmStringCreateLocalized("Another Button With no Menu");
  count = 0;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNx, 40); count++;
  XtSetArg(args[count], XmNy, 80); count++;
  btn3 = XmCreatePushButton(da2, "pb3", args, count);
  XtManageChild(btn3);

  count = 0;
  XtSetArg(args[count], XmNpopupEnabled, XmPOPUP_AUTOMATIC); count++;
  popup4 = XmCreatePopupMenu(btn1, "popupBtn1", args, count);

  addChildren(popup4, "SharedMenu");
  addButton(popup4, "Do something more");
  addButton(popup4, "What else?");


  str = XmStringCreateLocalized("PushMe");
  count = 0;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNx, 40); count++;
  XtSetArg(args[count], XmNy, 40); count++;
  btn2 = XmCreatePushButton(da1, "pb2", args, count);
  XtManageChild(btn2);

  str = XmStringCreateLocalized("I share a menu");
  count = 0;
  XtSetArg(args[count], XmNlabelString, str); count++;
  XtSetArg(args[count], XmNx, 240); count++;
  XtSetArg(args[count], XmNy, 40); count++;
  sbtn1 = XmCreatePushButton(da1, "sharepb1", args, count);
  XtManageChild(sbtn1);

  /* Associate a menu with this button */
  XmAddToPostFromList(popup4, sbtn1);

  count = 0;
  XtSetArg(args[count], XmNpopupEnabled, XmPOPUP_AUTOMATIC); count++;
  popup3 = XmCreatePopupMenu(btn2, "popupBtn2", args, count);

  addChildren(popup3, "Button2Menu");
  addButton(popup3, "Do something more");
  
  count = 0;
  XtSetArg(args[count], XmNpopupEnabled, True); count++;
  popup4 = XmCreatePopupMenu(da2, "PopupObjects1", args, count);

  addChildren(popup4, "ObjectMenu");
  addButton(popup4, "Flip");
  addButton(popup4, "Size");
  addButton(popup4, "Stretch");
  
  XtRealizeWidget(shell);

  XtAppMainLoop(context);

  return 0;    /* make compiler happy */
}


void 
QuitAppl(Widget w, char *i, XmPushButtonCallbackStruct *e)
{
  exit(0);
}

void 
addChildren(Widget popup, char* label)
{
  XmString str;
  Arg args[2];
  Widget child;

  str = XmStringCreateLocalized(label);
  XtSetArg(args[0], XmNlabelString, str);
  child = XmCreateLabelGadget(popup, "Title", args, 1);
  XtManageChild(child);
  XmStringFree(str);

  child = XmCreateSeparatorGadget(popup, "Sep", NULL, 0);
  XtManageChild(child);

  addButton(popup, "Cut");
  addButton(popup, "Copy");
  addButton(popup, "Paste");

}

void 
addButton(Widget parent, char* label) 
{
  Widget ch;
  XmString str;
  Arg args[2];

  str = XmStringCreateLocalized(label);
  XtSetArg(args[0], XmNlabelString, str);
  ch = XmCreatePushButtonGadget(parent, label, args, 1);
  XtManageChild(ch);
  XmStringFree(str);

}

GC drawGC = NULL;

void 
ExposeCB(Widget da, char *ig, XmDrawingAreaCallbackStruct *cbs)
{
  Arg args[2];
  XGCValues gcv;
  Pixel fg;
  Pixel bg;

  if (drawGC == NULL) {
    XtSetArg (args[0], XmNforeground, &fg);
    XtSetArg (args[1], XmNbackground, &bg);
    XtGetValues (da, args, 2);
    gcv.foreground = fg;
    gcv.background = bg;
    gcv.line_width = 2;
    drawGC = XtGetGC (da, GCForeground | GCBackground | GCLineWidth, &gcv);
  }

  XDrawArc(XtDisplay(da), XtWindow(da), drawGC,
	   circlex, circley, circled, circled, 
	   0, (360 * 64));
  XDrawRectangle(XtDisplay(da), XtWindow(da), drawGC, 
		 rectx, recty, rectw, recth);

}

void 
PopupHandlerCB(Widget da, XtPointer ignore, XmPopupHandlerCallbackStruct *phcs)
{
  int radius;
  int dx, dy;
  XButtonEvent *bevent = (XButtonEvent *) phcs -> event;
  int eventx, eventy;
  Position wox, woy;

  /* Translate event coordinates to this widget */
  /* First get origin of this widget */
  XtTranslateCoords(da, 0, 0, &wox, &woy);

  /* Now calculate relative offset */
  eventx = bevent -> x_root - wox;
  eventy = bevent -> y_root - woy;
  
  /* Check for hit in circle or rectangle to post special 
     menus */

  dx = eventx - rectx;
  dy = eventy - recty;

  if (dx > 0 && dx < rectw &&
      dy > 0 && dy < recth) {
    phcs -> menuToPost = popup4;
    return;
  }

  dx = eventx - circlex;
  dy = eventy - circley;

  dx -= circled / 2;
  dy -= circled / 2;

  radius = (dx * dx) + (dy * dy);

  if (radius < ((circled * circled) / 4))
    phcs -> menuToPost = popup4;
}
